src/update_handler.c

```diff
@@ -208,6 +238,106 @@ static gboolean ubifs_ioctl(RaucImage *image, int fd, GError **error)
 	return TRUE;
 }
 
+static gboolean splice_with_progress(GUnixInputStream *image_stream,
+		GUnixOutputStream *out_stream, goffset image_size, GError **error)
+{
+	ssize_t out_size = 0;
+	goffset sum_size = 0;
+	gint last_percent = -1, percent;
+
+	do {
+		int in_fd = g_unix_input_stream_get_fd(image_stream);
+		int out_fd = g_unix_output_stream_get_fd(out_stream);
+
+		/* Splice in 1M blocks */
+		out_size = splice(in_fd, NULL, out_fd, NULL, 1UL*1024*1024, 0);
+		if (out_size == -1) {
+			int err = errno;
+			g_set_error(error, G_IO_ERROR, g_io_error_from_errno(err),
+					"%s", g_strerror(err));
+			return FALSE;
+		}
+
+		sum_size += out_size;
+
+		percent = sum_size * 100 / image_size;
+		/* emit progress info (but only when in progress context) */
+		if (r_context()->progress && percent != last_percent) {
+			last_percent = percent;
+			r_context_set_step_percentage("copy_image", percent);
+		}
+	} while (out_size);
+
+	return TRUE;
+}
+
+static gboolean splice_file_to_outstream(const gchar *filename,
+		GUnixOutputStream *out_stream, GError **error)
+{
+	GError *ierror = NULL;
+	g_autoptr(GFile) image_file;
+	g_autoptr(GFileInfo) image_info;
+	goffset image_size;
+	g_autoptr(GUnixInputStream) image_stream = NULL;
+
+	g_return_val_if_fail(filename, FALSE);
+	g_return_val_if_fail(out_stream, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	image_file = g_file_new_for_path(filename);
+	image_info = g_file_query_info(image_file,
+			G_FILE_ATTRIBUTE_STANDARD_SIZE,
+			G_FILE_QUERY_INFO_NONE,
+			NULL,
+			&ierror);
+	if (image_info == NULL) {
+		g_propagate_error(error, ierror);
+		return FALSE;
+	}
+
+	image_size = g_file_info_get_size(image_info);
```

Comment:
This can be done with a `fstat()` in `splice_with_progress()`, which would simplify the logic.

Reply:
I am unsure how this simplifies the logic. With `fstat()` we still need a function call with error handling and having that inside `splice_with_progress()` will just move the code from one to another location.

Reply:
We don't need to open the file twice, as fstat simply works directly on the FD.